anesthetic/samples.py

```diff
@@ -389,13 +388,24 @@ class NestedSamples(MCMCSamples):
     beta: float
         thermodynamic temperature
 
+    logzero: float
+        The threshold for `log(0)` values assigned to rejected sample points.
+        Anything equal or below this value is set to `-numpy.inf`.
+        default: -numpy.inf
+
     """
 
     def __init__(self, *args, **kwargs):
         root = kwargs.pop('root', None)
+        logzero = kwargs.pop('logzero', -numpy.inf)
```

Comment:
I would probably have this as ``-1e30`` as a default, since that's what both MultiNest and PolyChord use.

Reply:
Better still would be to use the lowest value in `NestedSamples.logL_birth`

Reply:
> I would probably have this as `-1e30` as a default, since that's what both MultiNest and PolyChord use.

Would `<=-1e29` be safer to handle potential conversion issues?
I have spotted some `-0.99e30` values in the first column of a `_dead.txt` file. What does that column correspond to? Such values would be missed by a `-1e30` threshold but caught by a `-1e29` threshold...

> Better still would be to use the lowest value in `NestedSamples.logL_birth`

Is it certain that there is a `logzero` value in `logL_birth`? Wouldn't want to declare a physically meaningful value to be logzero...

Reply:
>I have spotted some -0.99e30 values in the first column of a _dead.txt file. What does that column correspond to? Such values would be missed by a -1e30 threshold but caught by a -1e29 threshold...

That's annoying. You could use a `numpy.isclose` test?

>Is it certain that there is a logzero value in logL_birth? Wouldn't want to declare a physically meaningful value to be logzero...

￼I would have thought so, although it seems like things can get garbled by the wrapper that's calling nested sampling (e.g. Cobaya)

I'm starting to wonder how useful this conversion is, and whether or not the modified logsumexp would be enough to fix the original issue in #71

Reply:
> That's annoying. You could use a `numpy.isclose` test?

I think `<=threshold` is more practical. That way I could specify `logzero=-1e30` as the threshold and it would work for multiple definitions, e.g. for both `-1e30` and `-1e300`. With `numpy.isclose` I would also have to pick some tolerance. I could simply apply that to the lower bound and test for `<=-1e29` or `<=-9e29`. Just a question of which tolerance to go for...?


> I'm starting to wonder how useful this conversion is, and whether or not the modified logsumexp would be enough to fix the original issue in #71

The new logsumexp currently tests against `-np.inf`. This only works if all logzero values have been converted. Otherwise `-1e300` values can still cause errors.

Also, part of the original problem was that `(-1e300)**2` was causing overflow errors completely independent of logsumexp.

Reply:
OK, agreed that we do need this functionality. The thresholding should be through an rtol, since this should only really occur with a floating-point type copying error. Perhaps at single precision (i.e. an rtol of 1e-8)?

Reply:
`rtol=1e-8` wouldn't even catch `-0.99e30`, so in that case I wouldn't bother with a `numpy.isclose`. 

Should I just set the default threshold to `-1e30` then and if needed it can be adjusted manually (athough as long as the really extreme values `-1e300` are taken care of, there shouldn't even be any problems anymore).

Reply:
Are the values `-0.99e30` or `-0.9999999999e30`? If it's the former, then it might be worth investigating if this is something that Cobaya is doing deliberately (say by increasing from logzero a little).

Reply:
It's `-0.99e30`.

Reply:
Ah, then this quite a specific cobaya issue. In ``cobaya.samplers.polychord.polychord.polychord.run`` it's return of `logpost` (equivalent to `loglike`) has the statement:

```python
 return max(logposterior + self.logvolume, 0.99 * self.pc_settings.logzero), derived
```

i.e. for some reason it's instructed to never return below logzero, which is actually quite counterproductive. @JesusTorrado could you shed some light on why cobaya puts in this factor of 0.99 ?

Reply:
I'm also not sure what the logvolume part is about. Nested sampling expects a likelihood, and the prior transformation in ``pc_prior`` should handle any normalisations like that. As written at the moment I think the log evidence would be incorrect.

Reply:
Ah, or perhaps it's undoing a volume transformation that is already done in the `model.logposterior` call, so that a likelihood is supplied (in which case this is the correct thing to do). That would make more sense. @JesusTorrado can you confirm?

Reply:
OK, I've opened an issue on [cobaya](https://github.com/CobayaSampler/cobaya/issues/77). As far as I can tell, Cobaya is effectively using -0.99e30 as its logzero value, so I believe the correct thing to do here is just to pass `logzero=-0.99e30` as the definition if the chains have come from cobaya.